Project1/ID.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+#include <utility>
+#include <map>
+#include <queue>
+
+template<typename T1, typename T2>
+class logic_pair :
+	public std::pair<T1, T2> {
+public:
+	using std::pair<T1, T2>::pair;
+	using std::pair<T1, T2>::operator=;
+	using std::pair<T1, T2>::swap;
+	using std::pair<T1, T2>::first;
+	using std::pair<T1, T2>::second;
+
+	bool operator==(const logic_pair& other) const {
+		return first == other.first && second == other.second;
+	}
+
+	bool operator!=(const logic_pair& other) const {
+		return first != other.first || second != other.second;
+	}
+
+	bool operator<(const logic_pair& other) const {
+		bool a = first < other.first;
+		bool b = second < other.second;
+		bool c = first == other.first;
+		return a | (c && b);
+	}
+	bool operator>(const logic_pair& other) const {
+		bool a = first > other.first;
+		bool b = second > other.second;
+		bool c = first == other.first;
+		return a | (c && b);
+	}
+
+	bool operator>=(const logic_pair& other) const {
+		bool a = first < other.first;
+		bool b = second < other.second;
+		bool c = first == other.first;
+		return !(a | (c && b));
+	}
+	bool operator<=(const logic_pair& other) const {
+		bool a = first > other.first;
+		bool b = second > other.second;
+		bool c = first == other.first;
+		return !(a | (c && b));
+	}
+
+};
+
+class ID :
```

Comment:
Is this not just:
```
class ID : private logic_pair<size_t, size_t> {
public:
    using logic_pair<size_t, size_t> operator==;
    using logic_pair<size_t, size_t> operator!=;
    using logic_pair<size_t, size_t> operator<=;
    using logic_pair<size_t, size_t> operator>=;
    using logic_pair<size_t, size_t> operator<;
    using logic_pair<size_t, size_t> operator>;
    friend class GridCell;
}
```

Reply:
I am sure I can find a way to use auto or decltype to fix this, but no. logic_pair< takes a logic_pair right operand, which this can't convert to.

Reply:
[Might want to take a look at this.](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern) I'm not sure if it solves the problem, but it looks like it'll be useful elsewhere if it isn't here.

I think you'd declare logic pair like:
```
template<typename A, typename B, typename Derived = logic_pair<A, B>>
class logic_pair {
    bool operator>=(const Derived& other) const {
        return whatever;
    }

```

Reply:
I tried that and it did not compile. Im kinda fine with how ID is right now, since I might make Grid and GridMap and ID have a template argument in the future to better support different map shapes, so ID may not always be a logic_pair anyway